kubewatcher/kubewatcher.go

```diff
@@ -258,50 +257,71 @@ func getResourceVersion(obj runtime.Object) (string, error) {
 func (ws *watchSubscription) eventDispatchLoop() {
 	log.Println("Listening to watch ", ws.Watch.Metadata.Name)
 	for {
-		for {
-			ev, more := <-ws.kubeWatch.ResultChan()
-			if !more {
+		// check watchSubscription is stopped or not before waiting for event
+		// comes from the kubeWatch.ResultChan(). This fix the edge case that
+		// new kubewatch is created in the restartWatch() while the old kubewatch
+		// is being used in watchSubscription.stop().
+		if ws.isStopped() {
+			ws.kubeWatch.Stop()
+			break
+		}
+		ev, more := <-ws.kubeWatch.ResultChan()
+		if !more {
+			if ws.isStopped() {
+				// watch is removed by user.
 				log.Println("Watch stopped", ws.Watch.Metadata.Name)
-				break
-			}
-			if ev.Type == watch.Error {
-				e := apierrs.FromObject(ev.Object)
-				log.Println("Watch error, retrying in a second: %v", e)
-				// Start from the beginning to get around "too old resource version"
-				ws.lastResourceVersion = ""
-				time.Sleep(time.Second)
-				break
-			}
-			rv, err := getResourceVersion(ev.Object)
-			if err != nil {
-				log.Printf("Error getting resourceVersion from object: %v", err)
+				return
 			} else {
-				log.Printf("rv=%v", rv)
-				ws.lastResourceVersion = rv
+				// watch closed due to unknown issues, restart it.
```

Comment:
Nitpick -- it's not an unknown issue, I just didn't know why when I first wrote this code :) The watch has a timeout, after which it returns and needs to be restarted. Would you mind changing this comment and log statement to something like "watch timed out, restarting"?